Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift

```diff
@@ -0,0 +1,162 @@
+
+    typealias ResponseContinuation = CheckedContinuation<URLResponse, any Error>
+    var responseContinuation: ResponseContinuation?
+
+    typealias ResponseBodyStream = AsyncBackpressuredStream<HTTPBody.ByteChunk, any Error>
+    var responseBodyStream: ResponseBodyStream
+    var responseBodyStreamSource: ResponseBodyStream.Source
+
+    /// Use `bidirectionalStreamingRequest(for:baseURL:requestBody:requestStreamBufferSize:responseStreamWatermarks:)`
+    init(requestBody: HTTPBody?, requestStreamBufferSize: Int, responseStreamWatermarks: (low: Int, high: Int)) {
+        self.requestBody = requestBody
+        self.hasAlreadyIteratedRequestBody = false
+        self.hasSuspendedURLSessionTask = false
+        self.requestStreamBufferSize = requestStreamBufferSize
+        (self.responseBodyStream, self.responseBodyStreamSource) = AsyncBackpressuredStream.makeStream(
+            backPressureStrategy: .highLowWatermarkWithElementCounts(
+                lowWatermark: responseStreamWatermarks.low,
+                highWatermark: responseStreamWatermarks.high
+            )
+        )
+    }
+
+    func urlSession(_ session: URLSession, needNewBodyStreamForTask task: URLSessionTask) async -> InputStream? {
+        debug("Task delegate: needNewBodyStreamForTask")
+        // If the HTTP body cannot be iterated multiple times then bad luck; the only thing
+        // we can do is cancel the task and return nil.
+        if hasAlreadyIteratedRequestBody {
+            guard requestBody!.iterationBehavior == .multiple else {
+                debug("Task delegate: Cannot rewind request body, canceling task")
+                task.cancel()
+                return nil
+            }
+        }
+
+        // Create a fresh pair of streams.
+        let (inputStream, outputStream) = createStreamPair(withBufferSize: requestStreamBufferSize)
+
+        // Bridge the output stream to the request body and open the output stream.
+        requestStream = HTTPBodyOutputStreamBridge(outputStream, requestBody!, openOutputStream: true)
+
+        // Return the new input stream (unopened, it gets opened by URLSession).
+        return inputStream
+    }
+
+    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
+        debug("Task delegate: didReceive data (numBytes: \(data.count), dataTask.state: \(dataTask.state))")
+        do {
+            switch try responseBodyStreamSource.write(contentsOf: [ArraySlice(data)]) {
+            case .produceMore: break
+            case .enqueueCallback(let writeToken):
+                if self.hasSuspendedURLSessionTask {
+                    debug("Task delegate: already suspended task, not enqueing another writer callback")
+                    break
+                }
+                debug("Task delegate: response stream backpressure, suspending URLSession task and enqueing callback")
+                dataTask.suspend()
+                self.hasSuspendedURLSessionTask = true
```

Comment:
What's the threading story here? Do we not need locks because the delegate calls are guaranteed to be called on a serial queue?

Reply:
My understanding was that delegate callbacks were called on the delegate queue.

However, I just put in some `dispatchPrecondition(condition: .onQueue(session.delegateQueue.underlyingQueue!))` and it appears not.

This is an important thing we need to get to the bottom of before merging this PR.

However, we should also consider what happens when this callback fires.